src/main/drivers/compass/compass_mpu925x_ak8963.c

```diff
@@ -0,0 +1,75 @@
+/*
+ * This file is part of Cleanflight.
+ *
+ * Cleanflight is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Cleanflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "platform.h"
+
+#include "drivers/bus.h"
+#include "drivers/sensor.h"
+#include "drivers/system.h"
+#include "drivers/time.h"
+
+#include "drivers/accgyro/accgyro_mpu.h"
+
+#include "drivers/compass/compass.h"
+#include "drivers/compass/compass_ak8963.h"
+#include "drivers/compass/compass_mpu925x_ak8963.h"
+
+#if defined(USE_MAG_MPU925X_AK8963)
+
+#define DETECTION_MAX_RETRY_COUNT   5
+
+#define MPU925X_I2C_ADDRESS             0x68
+#define AK8963_MAG_I2C_ADDRESS          0x0C
+#define MPU9250_BIT_RESET               0x80
+
+static bool mpu925xDeviceDetect(busDevice_t * dev)
+{
+    busWriteRegister(dev, MPU_RA_PWR_MGMT_1, MPU9250_BIT_RESET);
+    delay(150);
+    switch (busReadRegister(dev, MPU_RA_WHO_AM_I)) {
+        case MPU9250_WHO_AM_I_CONST:
+        case MPU9255_WHO_AM_I_CONST:
+            return true;
+        default:
+            return false;
+    }
+}
+
+bool mpu925Xak8963CompassDetect(magDev_t * mag)
+{
+    busDevice_t *busdev = &mag->busdev;
+    busdev->busdev_u.i2c.address = MPU925X_I2C_ADDRESS;
+    busDeviceRegister(busdev);
+    if (busdev == NULL) {
+        return false;
+    }
+    if (!mpu925xDeviceDetect(busdev)) {
+        return false;
+    }
+    // set bypass mode on mpu9250
+    busWriteRegister(busdev, MPU_RA_INT_PIN_CFG, 0x02);
```

Comment:
I've added code to reset the MPU if ak8963 detection fails and removed mpu925x_ak8963 from the default config generator. The mpu9250 can't be detected as gyro/acc over I2C, as there is no I2C driver for it.

Reply:
> The mpu9250 can't be detected as gyro/acc over I2C, as there is no I2C driver for it.

This is incorrect. The MPU9250 will be detected as a MPU6500 gyro/acc. The MPU9250 is a MPU6500 + AK8963 in one package.

Reply:
The code in drivers/accgyro/accgyro_mpu.c only checks the sig byte against MPU6500_WHO_AM_I_CONST, which is different from MPU9250_WHO_AM_I_CONST, so I assumed the mpu9250 couldn't be detected as mpu6500

Reply:
https://github.com/betaflight/betaflight/blob/1c3346bae14f0de9e7d37f566c96ed3eb8e1d226/src/main/drivers/accgyro/accgyro_mpu.c#L315

`MPU_INQUIRY_MASK` is 0x7E. MPU6500 identifier is 0x70, MPU9250 is 0x71. Masked with 0x7E they're both 0x70.

Reply:
I got a GY-91 board for testing, and as previously stated the MPU9250 is detected as a MPU6500 over I2C. Here is what the sensors detected - both gyro and BMP280 are detected as expected:
```
# status
MCU F40X Clock=168MHz (PLLP-HSE), Vref=3.28V, Core temp=50degC
Stack size: 2048, Stack address: 0x1000fff0
Configuration: CONFIGURED, size: 3991, max available: 16384
Devices detected: SPI:0, I2C:2
Gyros detected: gyro 1
GYRO=MPU6500, ACC=MPU6500, BARO=BMP280
```

Next I tested this PR when the gyro was configured to use I2C and while the AK8963 is detected and works, the gyro/acc (although detected earlier in the boot process) does not work properly. As a result of the mag detection and reconfiguring the device to be I2C passthrough it prevents the proper gyro/acc configuration from happening later in the boot process.

Since it would be moderately difficult to determine if a conflicting I2C gyro is detected before changing the MPU9250 to I2C passthrough, I would recommend changing the logic in `sensors/compass.c` for the function `compassDetect()` to not have the `case MAG_MPU925X_AK8963` be part of the `FALLTHROUGH` for auto-detect and instead require it to be explicitly set as `mag_hardware = MAG_MPU925X_AK8963`.

Reply:
Thanks for the help and suggestions! Just pushed a commit to address this